pyccel/ast/core.py

```diff
@@ -2537,6 +2577,108 @@ def __str__(self):
+                        else a for a in args)
+
+        # Replace the arguments in the code
+        self.body.substitute(self._orig_args+self.local_vars, self._new_args+self._new_local_vars, invalidate=False)
+
+    def swap_out_args(self):
+        """ Modify the body of the function by reinstating the original argument and local Variables
+        """
+        self.body.substitute(self._new_args+self._new_local_vars, self._orig_args+self.local_vars, invalidate=False)
+        self._new_args = None
+        self._new_local_vars = None
+
+    def remove_presence_checks(self):
+        """ Modify the body by replacing all expressions checking for the presence of an optional
+        variable. Either the If is removed or the check is replaced with its literal result
+        """
+        assert self._if_block_replacements is None
+        # Look for if blocks and replace present(x) statements
+        if_blocks = self.body.get_attribute_nodes(If, excluded_nodes=(FunctionDef,))
+        if_block_replacements = [[], []]
+        for i in if_blocks:
+            blocks = []
+            for c,e in i.blocks:
+                if isinstance(c, PyccelIs):
+                    if c.eval() is True:
+                        blocks.append((LiteralTrue(), e))
+                        break
+                    elif c.eval() is False:
+                        continue
+                else:
+                    presence_checks = c.get_attribute_nodes(PyccelIs, excluded_nodes = (FunctionDef,))
+                    for pi in presence_checks:
+                        if pi.eval() is True:
+                            if_block_replacements[0].append(pi)
+                            if_block_replacements[1].append(LiteralTrue())
+                        elif pi.eval() is False:
+                            if_block_replacements[0].append(pi)
+                            if_block_replacements[1].append(LiteralFalse())
+                blocks.append((c, e))
+            if len(blocks) == 0:
+                if_block_replacements[0].append(i)
+                if_block_replacements[1].append(EmptyNode())
+            elif len(blocks) == 1 and isinstance(blocks[0][0], LiteralTrue):
+                if_block_replacements[0].append(i)
+                if_block_replacements[1].append(blocks[0][1])
+            elif len(blocks) != len(i.blocks):
+                if_block_replacements[0].append(i)
+                if_block_replacements[1].append(If(*blocks))
+        self._if_block_replacements = if_block_replacements
+        self.body.substitute(if_block_replacements[0], if_block_replacements[1], invalidate=False)
+
+    def reinstate_presence_checks(self):
+        """ Modify the body by reinstating all expressions checking for the presence of an optional
+        variable
+        """
+        self.body.substitute(self._if_block_replacements[1], self._if_block_replacements[0])
+        self._if_block_replacements = None
```

Comment:
This method is used in the function `_handle_inline_func_call` in the fortran and c printers

Reply:
The function reverts the changes made by `remove_presence_checks` so that the `FunctionDef` can be called with different arguments

Reply:
I am afraid I don't understand why you make those changes and then revert them. One could keep the original function body and create a copy of it with the `presence` checks removed, right?

Reply:
There is currently no easy way to make a copy of a node and all its attributes (except `Variable`s and `FunctionDef`s) in pyccel. This issue can either be handled by adding a clone method to every class in pyccel (and I don't think this could be handled in `Basic`. I think each class would need to implement their own method). Or it can be handled by temporarily swapping out the arguments and presence checks in the body and then swapping them back in when the printing is finished

Reply:
Actually I haven't looked into this in detail, but I think you are right. Cloning can be complicated with our mutable objects, while it is much simpler in Sympy (which was probably designed with cloning in mind from its very beginning).

Reply:
In principle we could use `__slots__` and `setattr` to write a `clone` method for each of our classes. And maybe we would also need to use the `__slots__` of each of the parent classes. But does this cover all corner cases? 🤷‍♂️

Reply:
I don't think that will work unless you know a way to avoid calling `__init__`?

If you do, then it could work if we write 2 functions, `clone` for cloning the object and `clone_attribute` which would be called by `clone`. That way `clone_attribute` could be overridden for `FunctionDef` and `Variable` so they still point to the same object. We also need to avoid copying the old user/attribute nodes into the new objects

Reply:
Some `__new__` functions also specify the `args`/`kwargs`

Reply:
I do have a couple of ideas, but this topic is beyond the scope of this PR. We'll talk about it another time, possibly in a dedicated issue!